src/torchcodec/decoders/_core/VideoDecoder.cpp

```diff
@@ -846,7 +846,8 @@ VideoDecoder::RawDecodedOutput VideoDecoder::getDecodedOutputWithFilter(
 }
 
 VideoDecoder::DecodedOutput VideoDecoder::convertAVFrameToDecodedOutput(
-    VideoDecoder::RawDecodedOutput& rawOutput) {
+    VideoDecoder::RawDecodedOutput& rawOutput,
+    torch::Tensor& preAllocatedOutputTensor) {
```

Comment:
I dislike that we always have to pass a tensor. The current logic is that if an empty tensor was passed, then it's up to `convertAVFrameToDecodedOutputOnCPU` to do the allocation.

I tried `std::optional<torch::Tensor>& tensor = std::nullopt` but unfortunately, this requires the tensor to be `const` :(

I tried overloading, but that would lead to overloading a tons of different functions (the calling stack is long) and make things way more complex.

I tried passing a tensor pointer, but that make lifetime management impossible.

Do we have a better option than what is currently implemented?

Reply:
I think we do.

rawOutput itself has a pointer in it that callers populate. You can set the value of that instead of passing in a pre-allocated tensor as an extra parameter.

Reply:
The problem is that rawOutput isn't yet places like in `getFramesInRange`, where the output tensor is being pre-allocated. Can you confirm?

Reply:
A potential solution to this and the question of creating a bunch of empty tensors is to create a single empty tensor shared by the class:
```
Tensor EMPTY_TENSOR = torch::empty({0});
```
And then we could change this function definition to be:
```
VideoDecoder::DecodedOutput VideoDecoder::convertAVFrameToDecodedOutput(
    VideoDecoder::RawDecodedOutput& rawOutput,
    torch::Tensor& preAllocatedOutputTensor = &EMPTY_TENSOR) {
```
And then we do an identity-check on `EMPTY_TENSOR` in other places. We, of course, have to be very careful that we never modify `EMPTY_TENSOR`.

The other possibility is that we could just pass a pointer. That does not make lifetime management impossible, it just means we have to be very careful. In Rust terms, this would be a "borrow." The caller is responsible for the lifetime management of the object, but we are letting `convertAVFrameToDecodedOutput` _borrow_ the object to manipulate as it wants to. Once `convertAVFrameToDecodedOutput` returns back to us, we own the object again. In this world, the definition would be:
```
VideoDecoder::DecodedOutput VideoDecoder::convertAVFrameToDecodedOutput(
    VideoDecoder::RawDecodedOutput& rawOutput,
    torch::Tensor* preAllocatedOutputTensor = nullptr) {
```
And then we do a `nullptr` check to know if we have a preallocated tensor. I prefer the pointer approach, as the reference approach kinda nullifies (ha!) the point (ha!) of references, which is that they always refer to a real object.

If we were using [`shared_ptr`](https://en.cppreference.com/w/cpp/memory/shared_ptr) elsewhere, we would use that, but we're not. We really only make use of [`unique_ptr`](https://en.cppreference.com/w/cpp/memory/unique_ptr), which I think is fine.

Reply:
As discussed IMHO we should use std::optional<torch::Tensor> t = std::nullopt (note that there is no reference here).

It really is an input-only parameter, not an inout parameter so there is no need for a tensor.

The caller wont really look at this this parameter again -- it should only look at the return value.